BDMS-227-231-additional-well-info-transfer-scripts

ksmuczynski opened the pull request:
<!--
- BDMS-227-231-additional-well-info-transfer-scripts
-->
### **Why**
_This PR addresses the following problem / context:_
- Need to migrate legacy NM_Aquifer data to new Ocotillo database schema
- Legacy data has aquifer and formation information split across multiple lookup tables (LU_AquiferClass, LU_AquiferType, LU_Formations, LU_Lithology)
- Wells can have multiple aquifer characteristics (e.g., both "Fractured" and "Confined") which wasn't properly modeled
- Detailed stratigraphy data with depth intervals and lithology needs to be transferred from Stratigraphy.csv

The purpose of this PR is to make these intermediate updates available to @jacob-a-brown as we continue to co-work on the main BDMS-227 task.

### **How**
_Implementation summary - the following was changed / added / removed:_

**Transfer Scripts**
- Created `aquifer_system_transfer.py` to transfer aquifer systems from LU_AquiferClass (creates aquifer names with placeholder primary_type)
- Created `geologic_formation_transfer.py` to transfer formation codes from LU_Formations (stores only formation_code, ignores legacy MEANING values)
- Created `stratigraphy_transfer.py` to transfer detailed lithology logs from Stratigraphy.csv and create `ThingGeologicFormationAssociation` records (creates associations with depth intervals and updates formation lithology)
-  Updated `well_transfer.py` to:
    - Map `WellData.AqClass` codes to aquifer names using lexicon mapper
    - Parse multi-character `AquiferType` codes (e.g., "FC" → ["F", "C"])
    - Create `AquiferType` records for each characteristic
    - Update aquifer `primary_type` from "Unknown" to actual observed type
    - Handle missing `AqClass` by falling back to aquifer type name
    - Create `ThingAquiferAssociation` records linking wells to aquifer systems

**Utility Updates:**
- Updated `lexicon_mapper` to include `LU_AquiferClass` (CODE → aquifer name mapping)
- Updated `lexicon_mapper` to include `LU_Lithology` (ABBREVIATION → TERM mapping)
- Added special handling for `LU_Lithology` which uses ABBREVIATION/TERM instead of CODE/MEANING columns

**Transfer Workflow:**
1. `aquifer_system_transfer.py` - Creates named aquifer systems (e.g., "Ogallala Aquifer")
2. `geologic_formation_transfer.py` - Creates formation code records (e.g., "121ALVM")
3. `well_transfer.py` - Creates wells, aquifer associations, AquiferType records.
4. `stratigraphy_transfer.py` - Creates geologic formation associations with depth intervals and lithology

### **Notes**
_Any special considerations, workarounds, or follow-up work to note?_
**Follow-up work:**
- Create `formations.json` file with USGS formation code-to-name mappings for API
- Update API endpoints to map formation codes to names using formations.json
- May need to handle cases where wells have conflicting lithology data for same formation code

**Design Decisions:**
- Formation names intentionally NOT stored in database; API will map formation codes to names using authoritative formations.json file that can be updated independently
- `GeologicFormation.lithology` populated during stratigraphy transfer, not formation transfer, since lithology varies by location

ksmuczynski committed:
feat(transfers): add `LU_AquiferClass` and `LU_AquiferType` lookup tables to `util.py`.

Maps the `LU_AquiferClass` and `LU_AquiferType` lookup tables to the lexicon.

ksmuczynski committed:
feat(schemas): implement Create schemas for `aquifer_system` and `geologic_formation`.

ksmuczynski committed:
Merge branch 'kas-227-231-additional-well-info-models' into kas-227-231-additional-well-info-transfer-scripts

# Conflicts:
#	schemas/aquifer_system.py

ksmuczynski committed:
feat(transfer): WIP aquifer and geology transfers

ksmuczynski committed:
Merge branch 'bdms-227' into kas-227-231-additional-well-info-transfer-scripts

ksmuczynski committed:
feat(transfer): add stratigraphy transfer script and update geologic formation transfers.

Add stratigraphy_transfer.py to handle detailed lithology log import, create well-formation associations, and update formation lithology fields from stratigraphy data. This script is essential for linking wells to geologic formations with depth intervals.

jirhiker commented on transfers/aquifer_system_transfer.py:
```diff
@@ -0,0 +1,139 @@
+            break
+
+        # Log progress every 'step' rows
+        if i and not i % 25:
```
Replace 25 with step

jirhiker commented on transfers/stratigraphy_transfer.py:
```diff
@@ -0,0 +1,251 @@
+                top_depth = float(row.StratTop)
+                bottom_depth = float(row.StratBottom)
+            except (ValueError, TypeError) as e:
+                logger.warning(
```
Should these be critical log messages and recorded as an error that needs fixing

jirhiker reviewed (commented):
Looks good. A few minor tweaks needed

jacob-a-brown commented on schemas/geologic_formation.py:
```diff
@@ -5,6 +5,19 @@
+    Used during data transfer and API creation.
+    """
+
+    formation_code: str | None = None
```
because this is a restricted vocabulary you can use the enum created from the lexicon here

jacob-a-brown commented on schemas/geologic_formation.py:
```diff
@@ -5,6 +5,19 @@
+
+    formation_code: str | None = None
+    description: str | None = None
+    lithology: str | None = None
```
because this is a restricted vocabulary you can use the enum created from the lexicon here

jacob-a-brown commented on transfers/aquifer_system_transfer.py:
```diff
@@ -0,0 +1,139 @@
+        aquifer_name = row.MEANING
+
+        if not aquifer_name:
+            logger.warning(
```
this should be `logger.critical`

jacob-a-brown commented on transfers/stratigraphy_transfer.py:
```diff
@@ -0,0 +1,251 @@
+            # Validate required fields
+            # UnitIdentifier
+            if not hasattr(row, "UnitIdentifier") or not row.UnitIdentifier:
+                logger.warning(
```
should this be `logger.critical`?

ksmuczynski replied:
I think `.warning` is okay since these are related to expected data quality issues (not every Thing/well is going to have lithology or formation info). If they were unexpected, then I'd consider bumping them up a level.

jacob-a-brown commented on transfers/stratigraphy_transfer.py:
There's a bunch of validation being done here. Rather than spelling it all out here, and then having to redo the work later on, you can validate the data in a Pydantic model (and then use the same strategy of `model_validate`). This will keep the code cleaner and allow reuse later on.

ksmuczynski replied:
Do you think stratigraphy needs its own Pydantic schema? It does not have its own model. Maybe that doesn't matter? Is this worth fixing now or later?

jacob-a-brown replied:
If it is only going to be used in the data transfer then I don't think that it needs its own model. If, however, you think it'll be used again then I think it couldn't hurt (will that data need to be validated when adding new geologic data or associations?). But if that's not needed for a while it may not be worth the effort right now, so probably isn't needed at the moment

jacob-a-brown reviewed (commented):
Overall I think it looks good! Just left a few comments

ksmuczynski committed:
refactor(transfer): update log levels and error tracking in stratigraphy transfer

- Changed log level from 'warning' to 'critical' in the depth validation section
- Added error tracking and clearer error messages.

ksmuczynski committed:
refactor(schemas): use enums in `geologic_formation` create schema.

ksmuczynski committed:
refactor(transfer): update log levels and error tracking in aquifer_system transfer

- Changed log level from 'warning' to 'critical' in the depth validation section
- Added error tracking

ksmuczynski committed:
feat(transfer): Add aquifer system and geologic formation associations to well transfer

- Implement _extract_aquifer_type_codes() to parse compound codes (e.g., "FC" -> Fractured + Confined)
- Add get_or_create_aquifer() helper to manage unique aquifer system records
- Add get_or_create_formation() helper to manage geologic formation records
- Integrate aquifer association logic in `transfer_wells()` to create ThingAquiferAssociation and AquiferType records
- Integrate formation association logic to create ThingGeologicFormationAssociation records with depth data
- Support lexicon mapping for both AqClass (aquifer name) and AquiferType (characteristics) fields
- Add comprehensive error handling and logging for aquifer/formation associations

This enables proper tracking of wells' aquifer systems with multiple type characteristics
and their associated geologic formations, preserving all source data from NM_Aquifer.

ksmuczynski commented:
Made some updates to `well_transfer.py`:
- Implemented `_extract_aquifer_type_codes()` function to parse compound aquifer type codes (e.g., "FC" splits into ["F", "C"] for Fractured and Confined)
- Implemented `get_or_create_aquifer()` helper function that creates one aquifer system record per named aquifer (e.g., "Santa Fe Group") with deduplication logic
- Implemented `get_or_create_formation()` helper function to manage geologic formation records with deduplication
- Integrated aquifer association logic in `transfer_wells()` that:
  - Maps AqClass codes to aquifer names using lexicon mapper
  - Creates `ThingAquiferAssociation` records linking wells to aquifer systems
  - Creates separate `AquiferType` records for each characteristic (supports wells with multiple types)
  - Checks for existing associations to prevent duplicates
- Integrated formation association logic that:
  - Creates `ThingGeologicFormationAssociation` records with depth intervals

jacob-a-brown commented on transfers/well_transfer.py:
```diff
@@ -330,6 +436,131 @@ def transfer_wells(session: Session, flags: dict = None, limit: int = 0) -> None
+                            f"LU_AquiferType:{aquifer_codes[0]}"
+                        )
+
+                    # Determine primary type
```
this assumes that the first recorded type is the primary type of the aquifer. We should add a note here to ask AMMP (and get in touch with them)

jacob-a-brown commented on transfers/well_transfer.py:
```diff
@@ -330,6 +436,131 @@ def transfer_wells(session: Session, flags: dict = None, limit: int = 0) -> None
+                        )
+                        primary_type = None
+
+                    if primary_type:
```
if the primary type is not known can we still create an `aquifer_system` record and just set the primary type to `Unknown` like what is done in aquifer_system_transfer.py?

ksmuczynski replied:
Yes, that's a great idea.

jacob-a-brown commented on transfers/well_transfer.py:
```diff
@@ -330,6 +436,131 @@ def transfer_wells(session: Session, flags: dict = None, limit: int = 0) -> None
+                        row.WellDepth
+                        if row.WellDepth and not isna(row.WellDepth)
+                        else 100.0
+                    )
```
I don't think that these top and bottom depths should be assumed. if that data doesn't exist then I think it should be skipped, logged, and we can let AMMP handle it.

ksmuczynski replied:
I think top depth has to be assumed here. This is creating a single formation association from the WellData table, which only stores well depth (top depth is assumed 0.0 because it's at ground level). Detailed formation info with depth intervals is handled in `transfer_strigraphy.py`.

ksmuczynski committed:
refactor(transfer): Update logic related to creating an aquifer_system

- Set primary_type placeholder to "Unknown" instead of None when creating an aquifer_system in `well_transfer.py`

ksmuczynski committed:
refactor(transfer): Update logic related to creating formation associations

- Updated the logic so that a formation association is created only if valid well depth data exist.

ksmuczynski committed:
refactor(transfer): add note to verify compound aquifer type codes with AMMP

It is assumed that the first recorded type of a compound type is the primary type of the aquifer.

jacob-a-brown commented on transfers/well_transfer.py:
```diff
@@ -533,18 +536,28 @@ def transfer_wells(session: Session, flags: dict = None, limit: int = 0) -> None
-                        if row.WellDepth and not isna(row.WellDepth)
-                        else 100.0
-                    )
+                    # Onlyl create association if valid well depth data exists
```
good validation to check 👍

ksmuczynski committed:
refactor(transfer): removed unnecessary if statement re: creating aquifers

ksmuczynski committed:
Merge branch 'bdms-227' into kas-227-231-additional-well-info-transfer-scripts

# Conflicts:
#	transfers/well_transfer.py

jacob-a-brown commented on transfers/well_transfer.py:
```diff
@@ -355,6 +461,142 @@ def transfer_wells(session: Session, flags: dict = None, limit: int = 0) -> None
+                        geologic_formation=formation,
+                        top_depth=top_depth,
+                        bottom_depth=bottom_depth,
+                    )
```
An issue that comes to mind (and we just discussed) is: what should be done if the well record has a formation zone but doesn't have anything in the stratigraphy table? this will create associations indicating that the full well is that zone (top depth = 0 and bottom depth = well depth). perhaps we can store **WellData.FormationZone** in a field called `nma_formation_zone`? This will ensure the strictness of Ocotillo but also preserve legacy data if/when it is wanted.

ksmuczynski replied:
This is a thoughtful comment. I've added a `formation_completion_code` field to the Thing model and I've updated this section so it simply records which formation the well was completed in. `ThingGeologicFormationAssociation` records are no longer being created via the well transfer, they are only being created via the stratigraphy transfer.

jacob-a-brown replied:
Should the `formation_completion_code` be included in the `geologic_formations` field in the response? That is, added to the list of geologic formations a well passes through? Or should it be its own field to indicate this is where the well was completed?

On that note, @chasetmartin the step in the feature file says "And the response should include the formation as the formation zone of well completion." Should we return _all_ geologic formations associated with the well, or just the formation in which the well was completed?

jacob-a-brown replied:
If we are just returning the formation in which the well terminates I the following steps will need to be taken:

1. update the response schema
2. update the testing data
3. update the test for `the response should include the formation as the formation zone of well completion`

ksmuczynski replied:
I think the `formation_completion_code` should be its own field. I think there is value in distinguishing between what formation a well is completed in and what formations the borehole passes through.

chasetmartin replied:
@jacob-a-brown @ksmuczynski So just so I understand the question, there are multiple formations associated with a well in NMAquifer?

jacob-a-brown replied:
they're in the stratigraphy table (though there's also the field FormationZone for where the well terminates)

chasetmartin replied:
Ah yeah ok, the feature file was just referencing that FormationZone termination field, so Kelsey's suggestion above seems like a good plan.

ksmuczynski committed:
refactor(transfer): remove section that creates Formation Associations in `well_transfer.py`

The current implementation for creating a Formation Association was problematic for the following reasons:
	1. The FormationZone field from the WellData csv indicates completion zone, not the entire well stratigraphy
	2. Setting top_depth = 0.0 incorrectly implies:
		* The formation starts at ground surface
		* The well only penetrates one formation
		* The entire well depth is within that single formation
	3. ThingGeologicFormationAssociation = currently implies full stratigraphic column with depth intervals
	4. Forcing FormationZone into a depth-based association creates misleading data

Implementation
- Added `formation_completion_code` Field to Thing Model
	- This provides a clear separation between `formation_completion_code` = "What formation is the well completed in" and `formation_associations` = "What formations does the borehole pass through?"
- Updated well_transfer.py so that `ThingGeologicFormationAssociation` records are only being created from the Stratigraphy.csv (they were previously being created from WellData, too).

ksmuczynski committed:
refactor(transfer): add validation when assigning `formation_completion_code` to a well.

ksmuczynski committed:
refactor(schema): add `formation_completion_code` field to CreateWell schema.

jacob-a-brown commented on schemas/thing.py:
```diff
@@ -141,6 +141,7 @@ class CreateWell(CreateBaseThing, ValidateWell):
     well_construction_method_source: str | None = None
     well_pump_type: WellPumpType | None = None
     is_suitable_for_datalogger: bool | None
+    formation_completion_code: str | None = None
```
the `FormationCode` enum can be used here

jacob-a-brown reviewed (commented):
I think that it looks good - I just have one minor comment. The last thing to do is add the new transfer scripts to **transfers/transfer.py** and make sure the transfers are working (and do some spot checking between Ocotillo and NM_Aquifer to make sure the data is being transferred and persisted as expected)

ksmuczynski committed:
refactor(schema): use `FormationCode` enum for `formation_completion_code`

ksmuczynski committed:
refactor(transfer): add missing transfers to `transfer.py`

Added aquifer systems, geologic formations, and stratigraphy transfers.
